model_signing/_manifest.py

```diff
@@ -0,0 +1,101 @@
+    def from_string(s: str):
+        return DigestAlgorithm[s.replace("-", "_")]
+
+class Hashed:
+    algorithm: DigestAlgorithm
+    digest: bytes
+    def __init__(self, algorithm_: DigestAlgorithm, digest_: bytes):
+        self.algorithm = algorithm_
+        self.digest = digest_
+
+class PathMetadata:
+    hashed: Hashed
+    path: str
+    def __init__(self, path_: str, hashed_: Hashed):
+        self.path = path_
+        self.hashed = hashed_
+
+class Manifest:
+    paths: PathMetadata
+    predicate_type: str
+    def __init__(self, paths: [PathMetadata]):
+        self.paths = paths
+        self.predicate_type = "sigstore.dev/model-transparency/manifest/v1"   
+
+    def verify(self, verified_manifest: any) -> None:
+        # The manifest is the one constructed from disk and is untrusted.
+        # The statement is from the verified bundle and is trusted.
+        # Verify the type and version.
+        predicateType = verified_manifest["predicateType"]
+        if predicateType != self.predicate_type:
+            raise ValueError(f"invalid predicate type: {predicateType}")
+        files = verified_manifest["predicate"]["files"]
+        if len(self.paths) != len(files):
+            raise ValueError(f"mismatch number of files: expected {len(files)}, got {len(self.paths)}")
+        for i in range(len(self.paths)):
+            actual_path = self.paths[i]
+            verified_path = files[i]
+            # Verify the path.
+            if actual_path.path != verified_path["path"]:
+                raise ValueError(f"mismatch path name: expected '{verified_path['path']}'. Got '{actual_path.path}'")
+            # Verify the hash name in verified manifest.
+            if str(DigestAlgorithm.SHA256_P1) not in verified_path["digest"]:
+                raise ValueError(f"unrecognized hash algorithm: {set(verified_path['digest'].keys())}")
+            # Verify the hash name in actual path.
+            if actual_path.hashed.algorithm != DigestAlgorithm.SHA256_P1:
+                raise ValueError(f"internal error: algorithm {str(actual_path.hashed.algorithm)}")
+            # Verify the hash value.
+            verified_digest = verified_path["digest"][str(actual_path.hashed.algorithm)]
+            if actual_path.hashed.digest.hex() != verified_digest:
+                raise ValueError(f"mismatch hash for file '{actual_path.path}': expected '{verified_digest}'. Got '{actual_path.hashed.digest.hex()}'")
+
+
+    def to_intoto_statement(self) -> dsse.Statement:
+        # See example at https://github.com/in-toto/attestation/blob/main/python/tests/test_statement.py.
+        files: [any] = []
+        for _, p in enumerate(self.paths):
+            f = {
+                "path": p.path,
+                "digest": {
+                    str(p.hashed.algorithm): p.hashed.digest.hex(),
+                },
+            }
+            files += [f]
+        stmt = (
+            dsse._StatementBuilder()
+            .subjects(
```

Comment:
Shouldn't the subjects be the files? 

My mental model had the statement subjects as the artifacts we want to sign so you get `subject[n].name = path` and `subject[n].digest = file_hash`. So we would only set the `predicate_type` field as `sigstore.dev/model-transparency/manifest/v1` and that describes everything we need to know about the subjects. Furthermore, the predicate could then hold metadata like `{"hash_algorithm": "abc"}`.

PLMK if there is something I don't get here.

See https://github.com/in-toto/attestation/blob/main/spec/v1/statement.md

Reply:
You're correct that's how intoto works, but there is some discussion here about whether we want to use intoto or not https://github.com/sigstore/model-transparency/issues/111

For this PR, I could not pack files in the subject due to some current limitation on the hash type accepted by sigstore-python library (they only accept known hashes, but w euse a parallel hash).